keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,414 @@
+
+This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.
+
+## Motivation
+
+Today kubernetes doesn't enable storage providers that are able to safely detach volumes (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.
+
+### Goals
+
+* Increase the availability of stateful workloads.
+* Automate self-healing for stateful workloads.
+* This proposal only target CSI volumes.
+
+### Non-Goals
+
+* Implement in-cluster logic to handle node/control plane partitioning. 
+* Enable detach for all the storage providers.
+* Existing in-tree volumes are not targeted by this proposal.
+
+## Proposal
+
+User stories:
+
+* As a cluster administrator I want my stateful workload to failover in case a node is partitioned from the control plane without any intervention when it's safe to do so.
+
+* As a developer I would like to rely on a self-healing platform for my stateful workloads.
+
+This KEP plans to introduce the following CSI capabilities as described in this [PR](https://github.com/container-storage-interface/spec/pull/477).
+
+* In the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), introduce `UNPUBLISH_FENCE` controller service capability and `FORCE_UNPUBLISH` node service capability.
+  * The `UNPUBLISH_FENCE` controller service capability indicates that the SP supports ControllerUnpublishVolume.fence field.
+  * The `FORCE_UNPUBLISH` node service capability indicates that the SP supports the NodeUnpublishVolume.force field. It also indicates that the SP supports the NodeUnstageVolume.force field if it also has the STAGE_UNSTAGE_VOLUME node service capability.
+
+* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+  // SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations. When set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking whether a volume
+  // can be detached safely
+  // +optional
+  SafeDetach *bool
+}
+
+```
+
+Existing logic:
+* When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
```

Comment:
I think that this flow places too much burden on CSI plugins. The CO knows that the node has problems. The next decision is if/when to "shoot the node", or to allow/wait for it to recover. If waiting for it to recover, the workloads should not be harmed and no state should be migrated. When the waiting game is over, it's time to wreck the node.

*Dismantling the node vs. "shooting" it*

Dismantling the node by making a ton of CSI calls, `O(n)` where `n` is the number of vols, is likely error prone and doesn't scale well (if there are a large number of vols on a node). If multiple nodes are to be dismantled, then it's an `O(m * n)` problem, `m` is number of nodes. Which is even worse. Multiply that by the number of CSI plugins on the node (that support fencing). And also consider that CSI is probably not the only "service provider" that may be interested in this kind of event - who else might want to be part of this flow?

Dismantling the node, as proposed here, also means that more taints are added and pods are evicted. Does eviction apply to *all* pods (even daemonset and/or static pods)? Is it possible to write a pod spec that ignores the quarantine state (if not, should it be)? If pods were using actual hardware/firmware on the host, how does k8s reliably clear dirty state? Other HA systems use power-cycling to clear such state - there's no concept of that in k8s node lifecycle though, right? The WIP CSI PR was trying to promise something along those lines, but w/o the authority to e.g. power cycle (or whatever the infra runtime equivalent is for some cluster) a node (because it's outside the scope of CSI) the PR is a bit hamstrung.

Shooting the node by power cycling it is brutal. And effective w/ respect to dirty firmware/hardware state. Maybe part of the process can still include tainting the node (like the "quarantine" proposed here), and attempting to signal workloads before the node goes down. Though I'd like to see more discussion about what's allowed to run vs. not in that case. I'd also like to see k8s get an API (if it doesn't already have one) that indicates "hey, I've just rebooted and am ready for workloads again - you can trust that there's no dirty state hiding out somewhere waiting to corrupt your workloads"; otherwise, how does k8s know that it's safe to restart workloads on a node that has re-emerged?

Shooting the node also requires k8s to deal w/ "death match" and other fun (read: important) things. See https://ourobengr.com/ha/

Fencing like this should not be the job of individual components (e.g. CSI drivers, generic device drivers, etc.) but instead should be handled higher in the k8s stack. And maybe even involves integration w/ cluster-management APIs (such as Cluster API, owned by SIG Cluster Lifecycle) so that e.g. `Machine` objects are "power cycled" (if they support it), or simply deleted (and then a Cluster API controller can bring up a replacement).

Reply:
I'm going to narrow down the scope of this KEP and only handle the case when the node is really shutdown, leveraging the work from this KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown